bug: flake: FAIL: pkg/syncer/spec TestSpecSyncerProcess/SpecSyncer_sync_to_downstream,_syncer_finalizer_already_there

### Describe the bug

```
 === FAIL: pkg/syncer/spec TestSpecSyncerProcess/SpecSyncer_sync_to_downstream,_syncer_finalizer_already_there (0.15s)
I0110 17:19:39.459535   36355 shared_informer.go:255] Waiting for caches to sync for kcp-ddsif-gvr-source
I0110 17:19:39.459605   36355 shared_informer.go:262] Caches are synced for kcp-ddsif-gvr-source
I0110 17:19:39.459658   36355 shared_informer.go:255] Waiting for caches to sync for kcp-ddsif-gvr-source
I0110 17:19:39.459742   36355 shared_informer.go:262] Caches are synced for kcp-ddsif-gvr-source
    spec_process_test.go:1178: TestSpecSyncerProcess/SpecSyncer_sync_to_downstream,_syncer_finalizer_already_there: upstream ddsif synced
    spec_process_test.go:1180: TestSpecSyncerProcess/SpecSyncer_sync_to_downstream,_syncer_finalizer_already_there: downstream ddsif synced
    spec_process_test.go:1207: TestSpecSyncerProcess/SpecSyncer_sync_to_downstream,_syncer_finalizer_already_there: secrets and namespaces informers synced
    spec_process_test.go:1194: TestSpecSyncerProcess/SpecSyncer_sync_to_downstream,_syncer_finalizer_already_there: gvrs updated
I0110 17:19:39.462162   36355 spec_process.go:277] "Created downstream namespace for upstream namespace"
I0110 17:19:39.475732   36355 spec_process.go:519] "Upserted upstream resource to downstream" downstream.name="theDeployment"
    spec_process_test.go:1225: 
        	Error Trace:	spec_process_test.go:1225
        	Error:      	Should be empty, but was   []testing.Action{
        	            	  	testing.CreateActionImpl{ActionImpl: {Verb: "create", Resource: {Version: "v1", Resource: "namespaces"}}, Object: &unstructured.Unstructured{Object: {"apiVersion": string("v1"), "kind": string("Namespace"), "metadata": map[string]any{"annotations": map[string]any{"kcp.io/namespace-locator": string(`{"syncTarget":{"cluster":"root:org:ws","name":"us-west1","uid":"`...)}, "labels": map[string]any{"internal.workload.kcp.io/cluster": string("6ohB8yeXhwqTQVuBzJRgqcRJTpRjX7yTZu5g5g")}, "name": string("kcp-33jbiactwhg0")}}}},
        	            	+ 	testing.WatchActionImpl{
        	            	+ 		ActionImpl: testing.ActionImpl{Verb: "watch", Resource: s"apps/v1, Resource=deployments"},
        	            	+ 		WatchRestrictions: testing.WatchRestrictions{
        	            	+ 			Labels: s"internal.workload.kcp.io/cluster=6ohB8yeXhwqTQVuBzJRgqcRJTpRjX7y"...,
        	            	+ 			Fields: s"",
        	            	+ 		},
        	            	+ 	},
        	            	  	testing.PatchActionImpl{ActionImpl: {Namespace: "kcp-33jbiactwhg0", Verb: "patch", Resource: {Group: "apps", Version: "v1", Resource: "deployments"}}, Name: "theDeployment", PatchType: "application/apply-patch+yaml", Patch: `{"apiVersion":"apps/v1","kind":"Deployment","metadata":{"creatio`...},
        	            	  }
        	Test:       	TestSpecSyncerProcess/SpecSyncer_sync_to_downstream,_syncer_finalizer_already_there
    --- FAIL: TestSpecSyncerProcess/SpecSyncer_sync_to_downstream,_syncer_finalizer_already_there (0.15s)
=== FAIL: pkg/syncer/spec TestSpecSyncerProcess (2.06s) 
```

### Steps To Reproduce

.

### Expected Behaviour

.

### Additional Context

_No response_

--- a/pkg/syncer/spec/spec_process_test.go
+++ b/pkg/syncer/spec/spec_process_test.go
@@ -42,6 +42,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/apimachinery/pkg/watch"
 	dynamicfake "k8s.io/client-go/dynamic/fake"
 	"k8s.io/client-go/informers"
@@ -1135,20 +1136,6 @@ func TestSpecSyncerProcess(t *testing.T) {
 			)
 			require.NoError(t, err)
 
-			var cacheSyncsForAlwaysRequiredGVRs []cache.InformerSynced
-			for _, alwaysRequired := range []string{"secrets", "namespaces"} {
-				if informer, err := ddsifForUpstreamSyncer.ForResource(schema.GroupVersionResource{Group: "", Version: "v1", Resource: alwaysRequired}); err != nil {
-					require.NoError(t, err)
-				} else {
-					cacheSyncsForAlwaysRequiredGVRs = append(cacheSyncsForAlwaysRequiredGVRs, informer.Informer().HasSynced)
-				}
-				if informer, err := ddsifForDownstream.ForResource(schema.GroupVersionResource{Group: "", Version: "v1", Resource: alwaysRequired}); err != nil {
-					require.NoError(t, err)
-				} else {
-					cacheSyncsForAlwaysRequiredGVRs = append(cacheSyncsForAlwaysRequiredGVRs, informer.Informer().HasSynced)
-				}
-			}
-
 			setupServersideApplyPatchReactor(toClient)
 			resourceWatcherStarted := setupWatchReactor(tc.gvr.Resource, fromClusterClient)
 
@@ -1170,31 +1157,6 @@ func TestSpecSyncerProcess(t *testing.T) {
 			toInformerFactory.Start(ctx.Done())
 			toInformerFactory.WaitForCacheSync(ctx.Done())
 
-			gvrsUpdated := make(chan struct{})
-			upstreamSyncerDDSIFUpdated := ddsifForDownstream.Subscribe("upstreamSyncer")
-			downstreamDDSIFUpdated := ddsifForDownstream.Subscribe("downstream")
-			go func() {
-				<-upstreamSyncerDDSIFUpdated
-				t.Logf("%s: upstream ddsif synced", t.Name())
-				<-downstreamDDSIFUpdated
-				t.Logf("%s: downstream ddsif synced", t.Name())
-
-				_, unsynced := ddsifForUpstreamSyncer.Informers()
-				for _, gvr := range unsynced {
-					informer, _ := ddsifForUpstreamSyncer.ForResource(gvr)
-					cache.WaitForCacheSync(ctx.Done(), informer.Informer().HasSynced)
-					t.Logf("%s: upstream ddsif informer synced for gvr %s", t.Name(), gvr.String())
-				}
-				_, unsynced = ddsifForDownstream.Informers()
-				for _, gvr := range unsynced {
-					informer, _ := ddsifForDownstream.ForResource(gvr)
-					cache.WaitForCacheSync(ctx.Done(), informer.Informer().HasSynced)
-					t.Logf("%s: downstream ddsif informer synced for gvr %s", t.Name(), gvr.String())
-				}
-				t.Logf("%s: gvrs updated", t.Name())
-				gvrsUpdated <- struct{}{}
-			}()
-
 			ddsifForUpstreamSyncer.Start(ctx.Done())
 			ddsifForDownstream.Start(ctx.Done())
 
@@ -1203,11 +1165,32 @@ func TestSpecSyncerProcess(t *testing.T) {
 
 			<-resourceWatcherStarted
 
-			cache.WaitForCacheSync(ctx.Done(), cacheSyncsForAlwaysRequiredGVRs...)
-			t.Logf("%s: secrets and namespaces informers synced", t.Name())
+			// The only GVRs we care about are the 4 listed below
+			t.Logf("waiting for upstream and downstream dynamic informer factories to be synced")
+			gvrs := sets.NewString(
+				schema.GroupVersionResource{Group: "", Version: "v1", Resource: "namespaces"}.String(),
+				schema.GroupVersionResource{Group: "", Version: "v1", Resource: "configmaps"}.String(),
+				schema.GroupVersionResource{Group: "", Version: "v1", Resource: "secrets"}.String(),
+				schema.GroupVersionResource{Group: "apps", Version: "v1", Resource: "deployments"}.String(),
+			)
+			require.Eventually(t, func() bool {
+				syncedUpstream, _ := ddsifForUpstreamSyncer.Informers()
+				foundUpstream := sets.NewString()
+				for gvr := range syncedUpstream {
+					foundUpstream.Insert(gvr.String())
+				}
 
-			<-gvrsUpdated
+				syncedDownstream, _ := ddsifForDownstream.Informers()
+				foundDownstream := sets.NewString()
+				for gvr := range syncedDownstream {
+					foundDownstream.Insert(gvr.String())
+				}
+				return foundUpstream.IsSuperset(gvrs) && foundDownstream.IsSuperset(gvrs)
+			}, wait.ForeverTestTimeout, 100*time.Millisecond)
+			t.Logf("upstream and downstream dynamic informer factories are synced")
 
+			// Now that we know the informer factories have the GVRs we care about synced, we need to clear the
+			// actions so our expectations will be accurate.
 			fromClusterClient.ClearActions()
 			toClient.ClearActions()
 

